thunder/executors/nvfuserex_impl.py

```diff
@@ -2732,6 +2738,272 @@ def embedding(
 register_supported(ltorch.embedding, embedding, _embedding_check)
 
 
+def _cross_entropy_check_(
+    a: TensorLike,
+    /,
+    target: TensorLike,
+    weight: None | TensorLike,
+    size_average: None | Any,
+    ignore_index: int,
+    reduce: None | Any,
+    reduction: str,
+    label_smoothing: float,
+    *args,
+) -> bool:
+    if nvfuser_version() < LooseVersion("0.2.10"):
+        return False
+
+    if a.shape[-2] != target.shape[-1]:
+        return False
+
+    # input must be cast to float32
+    # since we use fmax which only supports float32
+    if dtypes.to_torch_dtype(a.dtype) != torch.float32:
+        return False
+
+    # We only optimize for the following cases
+    if reduction != "mean":
+        return False
+
+    if ignore_index >= 0:
+        return False
+
+    if (weight, size_average, reduce) != (None, None, None):
+        return False
+
+    if label_smoothing != 0.0:
+        return False
+
+    return True
+
+
+def cross_entropy_fwd_meta(
+    a: TensorLike,
+    /,
+    target: TensorLike,
+    weight: None | TensorLike = None,
+    size_average: None | Any = None,
+    ignore_index: int = -100,
+    reduce: None | Any = None,
+    reduction: str = "mean",
+    label_smoothing: float = 0.0,
+) -> tuple[TensorLike, TensorLike, TensorLike, TensorLike]:
+    losses: TensorLike
+    if reduction == "mean":
+        losses = TensorLike(like=a, shape=())
+    else:
+        check(
+            False,
+            lambda: f"cross entropy expected reduction to be  'mean' but was given {reduction}",
+        )
+
+    max_log_sum_exp: TensorLike = TensorLike(like=target, shape=target.shape, dtype=dtypes.float32)
+    num_valid_indices: TensorLike = TensorLike(like=losses, shape=(), dtype=dtypes.float32)
+    a_max: TensorLike = TensorLike(like=target, shape=target.shape, dtype=dtypes.float32)
+    return losses, a_max, max_log_sum_exp, num_valid_indices
+
+
+def cross_entropy_fwd(
+    a: TensorLike,
+    /,
+    target: TensorLike,
+    weight: None | TensorLike = None,
+    size_average: None | Any = None,
+    ignore_index: int = -100,
+    reduce: None | Any = None,
+    reduction: str = "mean",
+    label_smoothing: float = 0.0,
+    *,
+    fd: FusionDefinition,
+    lc_to_nv_map: dict,
+) -> Any:
+    nv_a = getnv(a, fd, lc_to_nv_map)
+    nv_target = getnv(target, fd, lc_to_nv_map)
+    nv_ignore_index = getnv(ignore_index, fd, lc_to_nv_map)
+
+    zero_scalar = fd.define_scalar(0, dtype=torch_dtype_to_nvfuser_dtype(dtypes.to_torch_dtype(a.dtype)))
+
+    # modify the labels to account for ignore index and then do a
+    # gather/ take_along_axis on the input tensor
+    ne = fd.ops.ne(nv_target, nv_ignore_index)
+    where_0 = fd.ops.where(ne, nv_target, zero_scalar)
```

Comment:
This feels a bit strange to me. For entries hitting ignore_index, we are clamping them to 0 on the index?

Shouldn't that require padding on the entry of `nv_a` in the first place?

Reply:
Just to clarify my question,

I think the proper behavior is to clamp the value after take_along_axis to 0, instead of clamping the target to zero. This felt like wrong math to me.

Reply:
For context: https://github.com/Lightning-AI/lightning-thunder/blob/4f35fcd2f674071d89a7145312dc5d672e616838/thunder/torch/__init__.py#L5562-L5588

Reply:
scratch this. I just missed the set-to-zero part later.